python/test/gluon/test_core.py

```diff
@@ -1432,3 +1436,620 @@ def kernel(out_ptr, M: ttgl.constexpr, N: ttgl.constexpr, K: ttgl.constexpr, a,
+
+    # Create values to scatter
+    values = torch.arange(N * M, dtype=torch.float32, device=device).reshape(N, M) + 100.0
+
+    output = torch.zeros((N, M), dtype=torch.float32, device=device)
+
+    # Expected: scatter values according to indices
+    expected = torch.zeros((N, M), dtype=torch.float32, device=device)
+    expected.scatter_(axis, indices.long(), values)
+
+    # Create layouts dynamically based on THREADS_PER_WARP
+    layout_2d = ttgl.BlockedLayout(size_per_thread=[1, 1], threads_per_warp=[THREADS_PER_WARP // 4, 4],
+                                   warps_per_cta=[1, 1], order=[1, 0])
+    shared_layout = ttgl.SwizzledSharedLayout(vec=1, per_phase=1, max_phase=1, order=[1, 0])
+
+    scatter_2d_kernel[(1, )](
+        indices,
+        values,
+        output,
+        N=N,
+        M=M,
+        axis=axis,
+        layout_2d=layout_2d,
+        shared_layout=shared_layout,
+        num_warps=1,
+    )
+
+    torch.testing.assert_close(output, expected)
+
+
+# ============================================================================
+# 3D Gather/Scatter Tests
+# ============================================================================
+
+
+@gluon.jit
+def gather_3d_kernel(
+    tensor_ptr,
+    indices_ptr,
+    output_ptr,
+    N: ttgl.constexpr,
+    M: ttgl.constexpr,
+    P: ttgl.constexpr,
+    axis: ttgl.constexpr,
+    layout_3d: ttgl.constexpr,
+    shared_layout: ttgl.constexpr,
+):
+    """Test 3D gather along specified axis."""
+    # Load the tensor from global memory [N, M, P]
+    # Create nested slice layouts: 3D -> 2D -> 1D
+    # For idx_n: [N,M,P] -> [N,P] -> [N]
+    slice_n_2d: ttgl.constexpr = ttgl.SliceLayout(dim=1, parent=layout_3d)  # Squeeze M
+    slice_n_1d: ttgl.constexpr = ttgl.SliceLayout(dim=1, parent=slice_n_2d)  # Squeeze P
+
+    # For idx_m: [N,M,P] -> [N,M] -> [M]
+    slice_m_2d: ttgl.constexpr = ttgl.SliceLayout(dim=2, parent=layout_3d)  # Squeeze P
+    slice_m_1d: ttgl.constexpr = ttgl.SliceLayout(dim=0, parent=slice_m_2d)  # Squeeze N
+
+    # For idx_p: [N,M,P] -> [N,P] -> [P]
+    slice_p_2d: ttgl.constexpr = ttgl.SliceLayout(dim=1, parent=layout_3d)  # Squeeze M
+    slice_p_1d: ttgl.constexpr = ttgl.SliceLayout(dim=0, parent=slice_p_2d)  # Squeeze N
+
+    idx_n = ttgl.arange(0, N, layout=slice_n_1d)
+    idx_n_2d_temp = ttgl.expand_dims(idx_n, 1)
+    idx_n_2d = ttgl.convert_layout(idx_n_2d_temp, slice_n_2d)
+    idx_n_3d = ttgl.expand_dims(idx_n_2d, 1)
+
+    idx_m = ttgl.arange(0, M, layout=slice_m_1d)
+    idx_m_2d_temp = ttgl.expand_dims(idx_m, 0)
+    idx_m_2d = ttgl.convert_layout(idx_m_2d_temp, slice_m_2d)
+    idx_m_3d = ttgl.expand_dims(idx_m_2d, 2)
+
+    idx_p = ttgl.arange(0, P, layout=slice_p_1d)
+    idx_p_2d_temp = ttgl.expand_dims(idx_p, 0)
+    idx_p_2d = ttgl.convert_layout(idx_p_2d_temp, slice_p_2d)
+    idx_p_3d = ttgl.expand_dims(idx_p_2d, 1)
+
+    offsets_3d = idx_n_3d * (M * P) + idx_m_3d * P + idx_p_3d
```

Comment:
This is screaming out for auto layout:

```suggestion
    idx_n = ttgl.arange(0, N)[:, None, None]
    idx_m = ttgl.arange(0, M)[None, :, None]
    idx_p = ttgl.arange(0, P)[None, None, :]

    offsets_3d = idx_n * (M * P) + idx_m * P + idx_p
    offsets_3d = ttgl.set_auto_layout(offsets_3d, layout_3d)
```

Reply:
Oh cool I didn't know about `set_auto_layout`. I thought the code I was writing for 3d slicing seemed overly complex but couldn't figure out how to do it simpler.